File: backend/models/user_prefs.py
"""
User Preferences Models (Phase 14).
Manages seller allowlists and blocklists for AutoBid control.
"""
from typing import List, Optional
from datetime import datetime
from sqlmodel import Field, SQLModel, JSON, Column
from sqlalchemy.dialects.postgresql import ARRAY, TEXT


class UserPreference(SQLModel, table=True):
    """
    User preferences for seller control.
    
    Manages per-user allowlist and blocklist for sellers.
    Integrated with AutoBid Engine for bid filtering.
    """
    __tablename__ = "user_preferences"
    
    id: int = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id", index=True, unique=True)
    team_id: int = Field(foreign_key="teams.id", index=True)
    
    # Seller lists (stored as JSON arrays)
    allowlist: List[str] = Field(
        default_factory=list,
        sa_column=Column(JSON, nullable=False, default=[])
    )
    blocklist: List[str] = Field(
        default_factory=list,
        sa_column=Column(JSON, nullable=False, default=[])
    )
    
    # Metadata
    notes: Optional[str] = Field(default=None, max_length=1000)
    updated_at: datetime = Field(default_factory=datetime.utcnow, index=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Config:
        arbitrary_types_allowed = True


class SellerItem(SQLModel):
    """Seller item for API requests/responses."""
    seller_id: str = Field(max_length=200, description="Seller identifier")
    source: str = Field(max_length=50, description="Marketplace source (ebay, etsy, etc.)")
    reason: Optional[str] = Field(default=None, max_length=500, description="Reason for preference")


class UserPreferencesResponse(SQLModel):
    """Response model for GET /api/user/prefs."""
    user_id: int
    team_id: int
    allowlist: List[SellerItem]
    blocklist: List[SellerItem]
    updated_at: str
    
    class Config:
        json_schema_extra = {
            "example": {
                "user_id": 1,
                "team_id": 1,
                "allowlist": [
                    {
                        "seller_id": "trusted-seller-001",
                        "source": "ebay",
                        "reason": "Highly reliable, consistent quality"
                    }
                ],
                "blocklist": [
                    {
                        "seller_id": "suspect-seller-042",
                        "source": "etsy",
                        "reason": "Low trust score, multiple failed transactions"
                    }
                ],
                "updated_at": "2025-11-03T12:00:00Z"
            }
        }


class UpdatePreferenceRequest(SQLModel):
    """Request model for POST /api/user/prefs/update."""
    action: str = Field(
        description="Action to perform",
        pattern="^(add_allow|remove_allow|add_block|remove_block)$"
    )
    seller_id: str = Field(max_length=200, description="Seller identifier")
    source: str = Field(max_length=50, description="Marketplace source")
    reason: Optional[str] = Field(default=None, max_length=500, description="Reason for preference")
    
    class Config:
        json_schema_extra = {
            "example": {
                "action": "add_block",
                "seller_id": "suspect-seller-042",
                "source": "etsy",
                "reason": "Low trust score, multiple failed transactions"
            }
        }


class UpdatePreferenceResponse(SQLModel):
    """Response model for POST /api/user/prefs/update."""
    success: bool
    message: str
    affected_list: str  # "allowlist" or "blocklist"
    seller_id: str
    new_total: int
    updated_at: str
    
    class Config:
        json_schema_extra = {
            "example": {
                "success": True,
                "message": "Seller added to blocklist",
                "affected_list": "blocklist",
                "seller_id": "suspect-seller-042",
                "new_total": 3,
                "updated_at": "2025-11-03T12:00:00Z"
            }
        }

File: frontend/src/api/userPrefs.ts
/**
 * User Preferences API Client (Phase 14)
 */
import axios from 'axios';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface SellerItem {
  seller_id: string;
  source: string;
  reason?: string | null;
}

export interface UserPreferencesResponse {
  user_id: number;
  team_id: number;
  allowlist: SellerItem[];
  blocklist: SellerItem[];
  updated_at: string;
}

export interface UpdatePreferenceRequest {
  action: 'add_allow' | 'remove_allow' | 'add_block' | 'remove_block';
  seller_id: string;
  source: string;
  reason?: string;
}

export interface UpdatePreferenceResponse {
  success: boolean;
  message: string;
  affected_list: string;
  seller_id: string;
  new_total: number;
  updated_at: string;
}

export interface CheckSellerResponse {
  allowed: boolean;
  reason: string;
}

/**
 * Get user's seller preferences
 */
export async function getUserPrefs(token: string): Promise<UserPreferencesResponse> {
  const response = await axios.get(`${API_BASE_URL}/api/user/prefs`, {
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });
  return response.data;
}

/**
 * Update user's seller preferences
 */
export async function updateUserPrefs(
  token: string,
  request: UpdatePreferenceRequest
): Promise<UpdatePreferenceResponse> {
  const response = await axios.post(
    `${API_BASE_URL}/api/user/prefs/update`,
    request,
    {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data;
}

/**
 * Check if a seller is allowed
 */
export async function checkSeller(
  token: string,
  seller_id: string,
  source: string
): Promise<CheckSellerResponse> {
  const response = await axios.get(
    `${API_BASE_URL}/api/user/prefs/check/${seller_id}`,
    {
      params: { source },
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    }
  );
  return response.data;
}
